File: app/src/lib/citations.ts
// Mirror of scripts/citations.py. Keep both in lockstep.

const ARRAY_FIELDS = ["evidence_record_ids", "record_ids"] as const;

const SINGLE_FIELDS = [
  "filing_id", "fppc_report_id", "form_700_line",
  "minutes_url", "agenda_url", "meeting_url",
  "docket_number", "permit_id",
  "source_filing_id", "fppc_id",
] as const;

const PAIR_REQUIRED: ReadonlyArray<readonly string[]> = [["source_url", "source_id"]];

function isSet(value: unknown): boolean {
  if (value == null) return false;
  if (typeof value === "string") return value.trim() !== "";
  if (Array.isArray(value)) return value.length > 0;
  return Boolean(value);
}

export function hasPrimarySourceCitation(node: Record<string, unknown>): boolean {
  for (const f of ARRAY_FIELDS) if (isSet(node[f])) return true;
  for (const f of SINGLE_FIELDS) if (isSet(node[f])) return true;
  for (const fields of PAIR_REQUIRED) {
    if (fields.every((f) => isSet(node[f]))) return true;
  }
  return false;
}

File: scripts/citations.py
"""has_primary_source_citation — node-level provenance check.

Used by adjacency-flow eligibility (spec §6.3.2) and ingestion validators.
Mirror of app/src/lib/citations.ts.
"""
from __future__ import annotations

ARRAY_FIELDS = ("evidence_record_ids", "record_ids")
SINGLE_FIELDS = (
    "filing_id", "fppc_report_id", "form_700_line",
    "minutes_url", "agenda_url", "meeting_url",
    "docket_number", "permit_id",
    "source_filing_id", "fppc_id",
)
PAIR_REQUIRED = (("source_url", "source_id"),)


def _is_set(value) -> bool:
    if value is None:
        return False
    if isinstance(value, str):
        return value.strip() != ""
    if isinstance(value, (list, tuple)):
        return len(value) > 0
    return bool(value)


def has_primary_source_citation(node: dict) -> bool:
    for f in ARRAY_FIELDS:
        if _is_set(node.get(f)):
            return True
    for f in SINGLE_FIELDS:
        if _is_set(node.get(f)):
            return True
    for fields in PAIR_REQUIRED:
        if all(_is_set(node.get(f)) for f in fields):
            return True
    return False
